### PR TITLE
PLA-1126 Feature/ad transfers

### PR DESCRIPTION
**Pull request by LorranSutter**

### CODE REVIEW

File: contracts/NFTs/AftermarketDeviceId.sol [hunk 1]
```diff
@@ -16,4 +19,35 @@ contract AftermarketDeviceId is Initializable, NftBaseUpgradeable {
...
+        address to,
+        uint256 tokenId
+    ) internal override {
+        // Approvals are not accepted for now
+        require(
+            msg.sender == address(_dimoRegistry) || msg.sender == from,
+            "Caller is not the owner"
+        );
+        _dimoRegistry.verifyAftermarketDeviceTransfer(tokenId);
+        ERC721Upgradeable._transfer(from, to, tokenId);
```
Comment (elffjs) [inline contracts/NFTs/AftermarketDeviceId.sol:51]: This is where inheritance gets confusing. It makes sense, though.

File: contracts/NFTs/AftermarketDeviceId.sol [hunk 2]
```diff
@@ -16,4 +19,35 @@ contract AftermarketDeviceId is Initializable, NftBaseUpgradeable {
...
+        address from,
+        address to,
+        uint256 tokenId
+    ) internal override {
+        // Approvals are not accepted for now
+        require(
+            msg.sender == address(_dimoRegistry) || msg.sender == from,
+            "Caller is not the owner"
+        );
+        _dimoRegistry.verifyAftermarketDeviceTransfer(tokenId);
```
Comment (elffjs) [inline contracts/NFTs/AftermarketDeviceId.sol:50]: Does this method on the registry seem overly specific to you? Asking whether—and to what—a vehicle is paired seems like something of general interest, even after we remove the transfer restriction.

I guess it's unclear who should "own" this guard.

File: contracts/NFTs/Base/NftBaseUpgradeable.sol [hunk 3]
```diff
@@ -9,8 +9,7 @@ import "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
 import "@openzeppelin/contracts-upgradeable/proxy/utils/UUPSUpgradeable.sol";
 import "@openzeppelin/contracts-upgradeable/utils/CountersUpgradeable.sol";
 
-/// @title MultiPrivilege
-/// @dev Based on the EIP-5496 https://eips.ethereum.org/EIPS/eip-5496
+/// @title NftBaseUpgradeable
```
Comment (elffjs) [inline contracts/NFTs/Base/NftBaseUpgradeable.sol:12]: 👍

File: contracts/NFTs/AftermarketDeviceId.sol [hunk 4]
```diff
@@ -16,4 +19,35 @@ contract AftermarketDeviceId is Initializable, NftBaseUpgradeable {
...
+    function setDimoRegistryAddress(address addr)
+        external
+        onlyRole(DEFAULT_ADMIN_ROLE)
+    {
+        require(addr != address(0), "Non zero address");
+        _dimoRegistry = IDimoRegistry(addr);
+    }
+
+    /// @notice Internal function to transfer a token
+    /// @dev Only the token owner can transfer (no approvals)
```
Comment (elffjs) [inline contracts/NFTs/AftermarketDeviceId.sol:35]: Good to note this 👍

File: test/AftermarketDeviceId.test.ts [hunk 5]
```diff
@@ -0,0 +1,403 @@
...
+            .connect(user1)['safeTransferFrom(address,address,uint256)'](user1.address, user2.address, 1)
+        ).to.be.revertedWith('AD must not be paired');
+      });
+    });
+
+    context('State', () => {
+      it('Should keep the same parent node', async () => {
+        const parentNode = await nodesInstance.getParentNode(adIdInstance.address, 2);
+
+        await adIdInstance.connect(user1)['safeTransferFrom(address,address,uint256)'](user1.address, user2.address, 2);
```
Comment (elffjs) [inline test/AftermarketDeviceId.test.ts:377]: There's some reason why you have to do this and for the life of me I can't remember what it is.

To be honest I wouldn't have thought to test these registry properties. This is good.

**Review by elffjs — APPROVED** [review]
I'm a little split on `verifyAftermarketDeviceTransfer`. Looks good overall.

File: test/AftermarketDeviceId.test.ts (hunk 5 above)
Comment (LorranSutter) [reply test/AftermarketDeviceId.test.ts:377]: Bc we have two `safeTransferFrom(address,address,uint256)` and `safeTransferFrom(address,address,uint256, bytes)`. So we cannot do simply `contract.safeTransferFrom`, we have to specify which one we want.

File: contracts/NFTs/AftermarketDeviceId.sol (hunk 2 above)
Comment (LorranSutter) [reply contracts/NFTs/AftermarketDeviceId.sol:50]: Humm... for now we just do the check actually in the `mapper` contract storage. But I think we might want to do more checks in the future of even change the state, so kinda makes sense this function be in the `AftermarketDevice`

Comment (elffjs) [reply contracts/NFTs/AftermarketDeviceId.sol:50]: I guess I'm asking whether one might prefer to just have something like `dimoRegistry.getVehiclePairedWithAftermarketDevice(tokenId) == 0`. "What vehicle is this connected to, if any?" seems like a natural on-chain question.

File: test/AftermarketDeviceId.test.ts (hunk 5 above)
Comment (elffjs) [reply test/AftermarketDeviceId.test.ts:377]: Oh yeah overloading ahahaha

File: contracts/NFTs/AftermarketDeviceId.sol (hunk 2 above)
Comment (elffjs) [reply contracts/NFTs/AftermarketDeviceId.sol:50]: Anyway, don't let this hold up getting this out there, I think it's good.

**Comment by elffjs** [comment]
Lorran is adding multi-privilege to this, so we will wait for that.

**Comment by elffjs** [comment]
I'm surprised that we need two new types—are these things not as orthogonal as I thought?

**Comment by LorranSutter** [comment]
> I'm surprised that we need two new types—are these things not as orthogonal as I thought?

I know, right. The thing is, I still need the `_transfer` from `MultiPrivilge`, `ERC721` and `AftermarketDeviceId`, bc each of them do something different. Also I removed `TRANSFERER_ROLE`, since it is not needed anymore. But I cannot use the same base contracts for `ManufacturerId` and `VehicleId` due to these specificities.

**Pull request merged**